persistent-postgresql/test/UpsertWhere.hs

```diff
@@ -0,0 +1,110 @@
+{-# LANGUAGE DataKinds, FlexibleInstances, MultiParamTypeClasses, ExistentialQuantification #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+module UpsertWhere where
+
+import Data.List              (sort)
+
+import Database.Persist.Postgresql
+import PgInit
+
+share [mkPersist sqlSettings, mkMigrate "upsertWhereMigrate"] [persistLowerCase|
+  Item
+     name        Text sqltype=varchar(80)
+     description Text
+     price       Double Maybe
+     quantity    Int Maybe
+
+     Primary name
+     deriving Eq Show Ord
+
+|]
+
+specs :: Spec
+specs = describe "UpsertWhere" $ do
+  let item1 = Item "item1" "" (Just 3) Nothing
+      item2 = Item "item2" "hello world" Nothing (Just 2)
+      items = [item1, item2]
+  describe "upsertWhere" $ do
+    it "inserts appropriately" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      upsertWhere item1 [ItemDescription =. "i am item 1"] []
+      Just item <- get (ItemKey "item1")
+      item @== item1
+
+    it "performs only updates given if record already exists" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      let newDescription = "I am a new description"
+      insert_ item1
+      upsertWhere
+        (Item "item1" "i am inserted description" (Just 1) (Just 2))
+        [ItemDescription =. newDescription] 
+        []
+      Just item <- get (ItemKey "item1")
+      item @== item1 { itemDescription = newDescription }
+
+  describe "upsertManyWhere" $ do
+    it "inserts fresh records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      let newItem = Item "item3" "fresh" Nothing Nothing
+      upsertManyWhere
+        (newItem : items)
+        [copyField ItemDescription]
+        []
+        []
+      dbItems <- map entityVal <$> selectList [] []
+      sort dbItems @== sort (newItem : items)
+    it "updates existing records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      upsertManyWhere
+        items
+        []
+        [ItemQuantity +=. Just 1]
+        []
+    -- TODO: this test doesn't pass
```

Comment:
this test is still failing in the following way
```
  1) UpsertWhere.upsertManyWhere only copies passing values
       expected: [Item {itemName = "item1", itemDescription = "", itemPrice = Just 6.0, itemQuantity = Nothing},Item {itemName = "item2", itemDescription = "hello world", itemPrice = Nothing, itemQuantity = Just 2}]
        but got: [Item {itemName = "item1", itemDescription = "", itemPrice = Just 6.0, itemQuantity = Nothing},Item {itemName = "item2", itemDescription = "hello world", itemPrice = Nothing, itemQuantity = Nothing}]
```

and here's the values + query
```
INSERT INTO "item"("name", "description", "price", "quantity") VALUES (?, ?, ?, ?), (?, ?, ?, ?) ON CONFLICT ("name") DO UPDATE SET "price"=EXCLUDED."price", "quantity"=COALESCE(NULLIF(EXCLUDED."quantity",?))
[PersistText "item1",PersistText "",PersistDouble 6.0,PersistInt64 0,PersistText "item2",PersistText "hello world",PersistNull,PersistInt64 0,PersistInt64 0]
```

any ideas as to what I could be missing?

Reply:
So, we expect the `item2 { itemQuantiuty = Just 2 }`, but it's actually `Nothing`. 

We've inserted `Just 2` as the value for the field in the database. The command we call is `upsertManyWhere newItems [ copyUnlessEq ItemQuantity (Just 0), .. ] [] []`, and for `newItems`, we set `itemQuantity = Just 0`. So, what we should expect here is that, since `Excluded.ItemQuantity == Just 0`, we shouldn't copy it over.

But, instead, we've got `NULL` in the database! So our `upsert` SQL is causing it to be `NULL` for some reason.